### PR TITLE
[WIP] Routing Prompt Widget

### LINKED ISSUE #73: Prompt widget for routing that renders a prompt before a route has transitioned
**Enhancement**

Add a new widget that can be used to render a view before a route transition has been completed based on a predicate. The render function would receive two functions that can be used to abort or continue the route transition.

For window reload/refresh, will need to use `onbeforeunload` and this only support the native browser confirm dialog.

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [ ] Unit or Functional tests are included in the PR

**Description:**

Provides a `Prompt` widget that can be used to prompt a user when they try to navigate away from a route and a the `shouldBlock` predicate returns `true`.

Resolves #73

### CODE REVIEW

File: src/routing/Prompt.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+			event.returnValue = true;
+		}
+	};
+
+	protected render(): DNode | DNode[] {
+		const item = this.registry.getInjector<Router>(this.properties.routerKey || 'router');
+		if (item) {
+			const router = item.injector();
+			if (!this._handle) {
+				this._handle = router.registerPrompt(this.properties.shouldBlock, () => {
```
Comment (JamesMilnerUK) [inline src/routing/Prompt.ts]: Would this benefit from being inside it's own function i.e. `this._addHandle`?

File: src/routing/Prompt.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+	private _onBeforeUnload = (event: Event) => {
+		if (this.properties.shouldBlock()) {
+			event.returnValue = true;
+		}
+	};
+
+	protected render(): DNode | DNode[] {
+		const item = this.registry.getInjector<Router>(this.properties.routerKey || 'router');
+		if (item) {
```
Comment (JamesMilnerUK) [inline src/routing/Prompt.ts]: Stylistic, but you could early return here to save an indent

File: src/routing/history/StateHistory.ts [hunk 3]
```diff
@@ -41,11 +42,21 @@ export class StateHistory implements HistoryInterface {
...
+		return this._previous;
+	}
+
+	public replace(path: string) {
+		const value = ensureLeadingSlash(path);
+		this._window.history.replaceState({}, '', this.prefix(value));
+	}
+
 	private _onChange = () => {
 		const value = stripBase(this._base, this._window.location.pathname + this._window.location.search);
```
Comment (JamesMilnerUK) [inline src/routing/history/StateHistory.ts]: Worth putting pathname + search into a variable?

**Comment by agubler** [comment]
Needs rebasing based on the changes to the router.

**Pull request closed without merging**